### ISSUE TITLE
Multilingual support in annotation

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#302]
**Modified by beutlich on 19 Dec 2016 08:25 UTC**
It would be nice if annotations texts in modelica specifications include any kind of multilingual support. 

Suppose a simple model with some annotations, every one in a different language:
```
model MyModel
  annotation(lang=en,Documentation(Info="<HTML>in english</HTML>"));
  annotation(lang=de,Documentation(Info="<HTML>in deutsch</HTML>"));
  annotation(lang=es,Documentation(Info="<HTML>in spanish</HTML>"));
...
end MyModel
```
modelica tools could show documentation in the language selected by the user.

Oscar Duarte

----

**Reported by ogduartev on 5 Mar 2010 14:17 UTC**
It would be nice if annotations texts in modelica specifications include any kind of multilingual support. 

Suppose a simple model with some annotations, every one in a different language:

model MyModel
  annotation(lang=en,Documentation(Info="<HTML>in english</HTML>"));
  annotation(lang=de,Documentation(Info="<HTML>in deutsch</HTML>"));
  annotation(lang=es,Documentation(Info="<HTML>in spanish</HTML>"));
...
end MyModel

modelica tools could show documentation in the language selected by the user.

Oscar Duarte

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/302

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 5 Mar 2010 15:01 UTC**
Oscar,

  This is an interesting idea.  I'm not sure if others have discussed this previously.  One comment on your proposed syntax.  I think it would be better to have the language more closely linked to the text.  For example, something like this:

```
model MyModel
    annotation(Documentation(lang="en",Info="<HTML>in English</HTML>"),
               Documentation(lang="de",Info="<HTML>auf Deutsch</HTML>"),
               Documentation(lang="es",Info="<HTML>en Español</HTML>"));
...
end MyModel;
```

Now we have a single annotation and it is clear that the text is associated directly with a language.  Of course, it is entirely possible that there is a proposal (or even a standard annotation) to handle this already.  I'm mostly commenting on how to make sure the coupling between language and text is explicit.

As far as I can tell, your suggestion would be totally backward compatible so long as the default lang was "en".

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 5 Mar 2010 15:09 UTC**
Unhiding so reporter can actually look at his ticket.

**Comment by modelica-trac-importer** [comment]
**Comment by PHarman on 24 May 2010 15:58 UTC**
Can I propose another alternative which may be "more" backwards compatible, and doesn't require multiple annotations of the same name:

```
model MyModel
    annotation(Documentation(info(de="<HTML>auf Deutsch</HTML>",es="<HTML>en Español</HTML>")="<HTML>in English</HTML>"));
...
end MyModel;
```

A German localised tool can then look for "Documentation.info.de" and if it doesn't find it revert to the default English version at "Documentation.info".

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 31 Mar 2016 08:15 UTC**
Replying to [comment:3 PHarman]:
> Can I propose another alternative which may be "more" backwards compatible, and doesn't require multiple annotations of the same name:
> 
> ```
> model MyModel
>     annotation(Documentation(info(de="<HTML>auf Deutsch</HTML>",es="<HTML>en Español</HTML>")="<HTML>in English</HTML>"));
> ...
> end MyModel;
> ```
> 
> A German localised tool can then look for "Documentation.info.de" and if it doesn't find it revert to the default English version at "Documentation.info".

I noticed this old ticket, and noticed two related problems with the proposal:
* It doesn't work for all strings - e.g. description strings are not included, and they are quite important for parameters.
* It isn't related to the normal work-process for translators when localizing soft-ware.

As far as I know most localization just uses a "data-base" of simple <name,value> pairs for strings (and possibly compression etc). Some things vary - quoting, extraction, special tools for changing strings - but in general people are not storing all localized variants together in the source.

Obviously we need to standardize on some variant of that - but the basic idea would just be a file "model-de" with (or preferably using an XML variant such as Qt's ts format or XLIFF instead - but I assume you get the idea):
```
MyModel.description=Unsere model
MyModel.info=<HTML>auf Deutsch</HTML>
```
(and obviously if the translation isn't found revert to the default string).

**Comment by modelica-trac-importer** [comment]
**Changelog removed by hansolsson on 31 Mar 2016 08:15 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by leo&#x2e;gall on 31 Mar 2016 09:33 UTC**
Replying to [comment:4 hansolsson]:
> As far as I know most localization just uses a "data-base" of simple <name,value> pairs for strings (and possibly compression etc). Some things vary - quoting, extraction, special tools for changing strings - but in general people are not storing all localized variants together in the source.
> 

This sounds like the right design for public/commercial libraries.

But, for custom libraries, many users (engineers) will work in their native language in Modelica.
So, if we have multi-lang support in tools (which would be great), there should be a way to declare the default language in the respective model/package.
This language definition holds for all description strings as well as Documentation annotations.

Otherwise we would loose the important ability of Modelica, to have one .mo file which does not depend on other files.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 31 Mar 2016 09:43 UTC**
There is another big benefit of having the translation files separate: it makes it possible to not parse Modelica text for the languages that you do not use (you will need at most one extra file). Usually the language stored in the main source-file is English (well... the C-locale) with other languages mirroring it.

Other benefits include the ability to add translations after a release, and not changing line numbers when adding new translations.

Translations are rather clear resources and not part of the actual model.

Note that one of the most common translation file formats (gettext) uses .po/.mo file extensions with only .po-files being stored in version control and .mo-files being generated from these when making a release (typically removing the .po-files for binary distributions). Qt is similar (.ts/.qm).

You typically get low overhead by using these binary formats as separate files.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 31 Mar 2016 10:14 UTC**
Replying to [comment:5 leo.gall]:
> Replying to [comment:4 hansolsson]:
> > As far as I know most localization just uses a "data-base" of simple <name,value> pairs for strings (and possibly compression etc). Some things vary - quoting, extraction, special tools for changing strings - but in general people are not storing all localized variants together in the source.
> > 
> 
> This sounds like the right design for public/commercial libraries.
> 
> But, for custom libraries, many users (engineers) will work in their native language in Modelica.
> So, if we have multi-lang support in tools (which would be great), there should be a way to declare the default language in the respective model/package.
> This language definition holds for all description strings as well as Documentation annotations.

I agree that it would be good, and having "default-language" in the library as you propose will improve the process, and just be clearer in general.

But it will not significantly alter the work-flow.

The idea is that you take the source library and generate a default translation file (based on the model contents) - and then translate that to different languages. The "default-language" of the model will basically just be propagated to the default translation file - and avoids a manual step (and requires you to find German->French translator instead of English->French translator).

For the user of a library: a translation is based on their preferences (where "default language in library" can be ordered properly).

If the model is written in German the default text will be German - and we can provide an English, Swedish, Japanese, French (and ...) translation. We could even provide a German translation different from the model (in case the engineers don't write proper German).

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 12 Dec 2016 11:01 UTC**
One other proposal is to add Comment.language (and Documentation.info.language) like
```mo
parameter Real var=1.0 "Default Comment"
  annotation(Comment(language="de")="Deutscher Kommentar (German Comment)");
```

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 19 Dec 2016 08:25 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by volker&#46;beuter on 10 Jan 2017 11:03 UTC**
Replying to [comment:8 beutlich]:
> One other proposal is to add Comment.language (and Documentation.info.language) like
> ```mo
> parameter Real var=1.0 "Default Comment"
>   annotation(Comment(language="de")="Deutscher Kommentar (German Comment)");
> ```

Couldn't this also been done in a translation file in the style like proposed by Hans:
```
MyModel.var.description="Deutscher Kommentar (German Comment)"
```
I think it ought to be consistent, not the translated documentation in a separate file (per language) but the description strings in an translation annotation.

But in case it turns out to be required to introduce such annotations, the selected language ought to be provided as a value (language="de"), not as a keyword (de = " ... ") in the style like proposed by Michael Tiller: If we include the language abbreviations into the annotation definition we either have to restrict to some 'big' languages (and so excluding some exotic translations) or we have to add a long list of all possibly ever used languages into the annotation definition.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 27 Jan 2017 07:56 UTC**
Hopefully we can find some agreement for MLS v3.4.

**Comment by modelica-trac-importer** [comment]
**Comment by richard&#x2e;haehne on 3 Feb 2017 12:43 UTC**
Replying to [comment:6 sjoelund.se]:
> There is another big benefit of having the translation files separate: it makes it possible to not parse Modelica text for the languages that you do not use (you will need at most one extra file). [...]
> 
> Other benefits include the ability to add translations after a release, and not changing line numbers when adding new translations. [...]
> 
> Note that one of the most common translation file formats (gettext) uses .po/.mo file extensions with only .po-files being stored in version control and .mo-files being generated from these when making a release (typically removing the .po-files for binary distributions). [...]

One way of using [GNU gettext](https://www.gnu.org/software/gettext/) is that you can have language independent identifier strings in the source files (here: the Modelica source). Those identifier strings can use simple english terms.

The idea of gettext is to have a template file between the source and the language, which eases translations a lot:
- The so called **Portable Object Template** (\*.pot) contains all of those identifier strings.
- The so called **Portable Object** (\*.po) uses that list of identifier strings to provide the translation. So each language has its independent \*.po file (and if necessary each language can be split into several \*.po files).

One advantage of this approach is:
> But, for custom libraries, many users (engineers) will work in their native language in Modelica.

When using english identifier strings, the engineer has to use little english only (for the identifier) and can use his/her mother tongue for the documentation itself.

Furthermore when deploying a Modelica text together with the Portable Object Template, other developers can easily add their own translation.

Disadvantage:
1. When the identifier in the source changes, then the Portable Object files have to be updated according to the Portable Object Template.
1. Modelica text has to be deployed together with at least one language file since identifiers are no _fallback language_.

(1) The update of the identifiers in the Portable Object Template and the Portable Objects could be automated when the sequence of the identifier strings stays the same.
(2) This could be neutralized when using the documentation itself as the identifier strings i.e. as _fallback language_. Since the language itself is used for the identifier string, then _technically_ it is not required to set a default language by the tool or Modelica text even if the engineer prefers to use his native language instead of english.

Workflow example for C (from Wikipedia):
![Image](https://upload.wikimedia.org/wikipedia/commons/6/6b/Gettext.svg)

GNU gettext was released in 1990 and meanwhile is supported by every major programming language and open source tools for editing portable object files are available (e.g. [POEdit](https://poedit.net/download)). From my humble point of view (being new to Modelica), Modelica should heed such a widely known workflow instead of reinventing the wheel ...

Replying to [comment:4 hansolsson]:
> but in general people are not storing all localized variants together in the source [...] preferably using an XML variant such as Qt's ts format or XLIFF

... and can use also different containers for the languages (where XLIFF might be prefered over Qt ts since [XLIFF is standardized](http://docs.oasis-open.org/xliff/xliff-core/v2.0/xliff-core-v2.0.html)).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 3 Feb 2017 15:03 UTC**
Replying to [comment:12 richard.haehne <richard.haehne@…>]:
> Replying to [comment:6 sjoelund.se]:
> > There is another big benefit of having the translation files separate: it makes it possible to not parse Modelica text for the languages that you do not use (you will need at most one extra file). [...]
> > 
> > Other benefits include the ability to add translations after a release, and not changing line numbers when adding new translations. [...]
> > 
> > Note that one of the most common translation file formats (gettext) uses .po/.mo file extensions with only .po-files being stored in version control and .mo-files being generated from these when making a release (typically removing the .po-files for binary distributions). [...]
> One way of using [GNU gettext](https://www.gnu.org/software/gettext/) is that you can have language independent identifier strings in the source files (here: the Modelica source). Those identifier strings can use simple english terms.

I don't think we should **add** identifier strings to the Modelica source - because most texts already have a unique identifier:

The description texts for variables and classes, and the info- and revisions-texts for classes have unique identifiers based on the path to that string (and some special symbol for info/revisions).

(There are some additional ones: texts in the graphical layer, description texts for choices-annotations. We may need identifiers for those, or just use the original texts as identifiers with some suitable prefix. Such identifiers could also help with localizing graphical symbols etc.)

Thus I think that no change of MSL is needed - at least not in the first iteration.

The issues are instead:
* Deciding on the exact mapping from Modelica to such identifiers.
* Deciding on the translation format.
* Implementing it.
* Actually translating the texts.

The last point is quite important - and expensive. I believe many are reluctant to commit to any action, since they understand that the last point will take a lot of effort.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 3 Feb 2017 15:10 UTC**
I just tried how easy it is to generate the POT file for the MSL. In case you are curious, see https://gist.github.com/tbeu/e971cbde840b61f179cfb57850e259bb for the 19k strings. The good thing is that existing libraries and also the Modelica standard would not have to be changed. We only would need to standardize where to search for the PO files and how to interprete them.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 10 Feb 2017 13:52 UTC**
I noticed that OMEdit displays trailing HTML text of comment strings as tooltips when hovering with the mouse on the component. Taking this idea, it would also be possible to add the localization strings as trailing XML, e.g.,
```mo
parameter Real var=1.0 "Default Comment<?xml version=\"1.0\"?><language><de>Deutscher Kommentar (German Comment)</de><language>"
```

One more thought, inline localization strings (as well as the HTML mouse tooltip) can be seen as one utilization of the "custom annotation" feature, discussed in #1142, #1145 and #1293. Thus both issues could be resolved in parallel, if the inline localization is to preferred over the global localization mentioned by comment:12.

Combining both ideas, custom annotations could also be designed by trailing XML text.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 28 Feb 2017 16:49 UTC**
Hans: My reasons for preferring external translation:
Size for parsing.
Parallel work on multiple translations.
Translators normally in other tools; and should be close to their work-flow.
Translation of encrypted libraries?

Should at least specify format and the keys for translations.
One possibility is to just use the string as key.
Advantage: Simple.
Disadvantage: Long keys, same text in different places must be translated in the same way, problem when updating texts.
Another is to have a fixed format: full-lexical-path (and some special symbols).

Modelica.Mechanics.Rotational.Components.Inertia.J.description: "Moment of inertia"
(can also translate package names?)

Should test these variants, but too late for Modelica 3.4.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 28 Feb 2017 16:53 UTC**
Poll on general direction:
External translation files: 8
Inside Modelica files: 0

Should check what formats translators work with:
QT translator framework.
XLiff
POT

**Comment by modelica-trac-importer** [comment]
**Comment by Tobias Zawada on 14 Sep 2017 10:09 UTC**
Most comments are written as comment strings. If we allowed constant string expressions as comment strings comments could already be localized.

The only requirement is that those expression must be evaluable within the GUI (i.e., without a full symbolic analysis).

Currently, SimulationX already has a `language` function with two string arguments. The first argument is the language identifier, e.g., `"en"`, `"de"`,... The second argument is a string that will be passed as return value if the first argument matches the current language otherwise an empty string will be returned. Up to now this function is only used in `assert`-statements.

If comment strings would be expressions one could already write:

```
parameter Real p=1 language("en","english comment")+
                   language("de","deutscher Kommentar");
```

Functions that retrieve comments from external files could also be applied. The decision is left to the user.

The standard could provide utility functions for the string generation that would lead to a greatly unified style in the libraries. But, there is no need to enforce one special approach by the standard.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 4 Oct 2017 10:44 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 20 Oct 2017 12:17 UTC**
Full context (lexical name) vs. just English text as names; conclusion seems to be use lexical name.

What to translate:
* Description (of component and classes)
* Choices - fully qualified name+choice-description as key, note that choice might be {1,0,0} or {0,-1,0} as in MBS
* Groups and tabs annotations (skip component-name and use group-name as key)
* Info (and revisions)-texts
* Texts in diagram: fully qualified name+text as key?

General conclusion: fully qualified name+string as key - skip string if not needed.

Problem with text in diagram is that some strings should not be translated, e.g. "%p".
This is in reality more complicated: "freqHz=%freqHz" and "amplitude" - to explain the parameter "amplitude".

Martin S: Mark texts in diagrams that should be translated, Qt has tr so textString=tr("...."). Could also use to translate other strings - like diagnostics for assertions.
Henrik: Python have prefixes for some strings.
Hans: Should ensure that has no current meaning.
Henrik: Easy to have different key for a string with tr - e.g. tr("...", "myKey");

One reason for duplicate strings are missing base-classes - investigate.

Should be straightforward to experiment with different variants of key-generation in a test-implementation.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 10 Jan 2018 08:23 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 6 Mar 2018 10:08 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 23 Mar 2018 08:05 UTC**

**Comment by beutlich** [comment]
Proof of concept with msgctxt was provided by SystemModeler.

Conclusion: Prepare MCP

**Comment by beutlich** [comment]
The test library and the proof of concept is now also available as separate repository, see https://github.com/ESI-ITI-GmbH/TranslationTest. @Daiwen @FanliZHOU @ooer

**Comment by beutlich** [comment]
We have now finished the prototype implementation for SimulationX.

Please note, that we defined the translation context for components to always be the scope of the enclosing type. For example, the package `P`

```mo
package P "Description"
  model M "Description"
    parameter Real p=0 "Description" annotation(Dialog(tab="Description", group="Description"));
    Real v=p "Description" annotation(Dialog(tab="Description", group="Description"));
    annotation(Documentation(info="Description"));
  end M;
end P;
```

would generate a POT file P.pot where all the strings of `P.M` have the same translation.

```pot
# Description of P
msgctxt "P"
msgid "Description"
msgstr ""

# Description of P.M, P.M.p, P.M.p.Dialog.tab, P.M.p.Dialog.group, P.M.v, P.M.v.Dialog.tab, P.M.v.Dialog.group, P.M.Documentation.info
msgctxt "P.M"
msgid "Description"
msgstr ""
```
